### PR TITLE
PSP-5431 : Project: Documents

### PR DESCRIPTION
**Pull request by eddherrera**
image

### CODE REVIEW

File: source/backend/api/Services/DocumentFileService.cs [hunk 1]
```diff
@@ -135,6 +170,24 @@ public async Task<ExternalResult<string>> DeleteResearchDocumentAsync(PimsResear
...
+        public async Task<ExternalResult<string>> DeleteProjectDocumentAsync(PimsProjectDocument projectDocument)
+        {
+            this.Logger.LogInformation("Deleting PIMS document for single Project");
+            this.User.ThrowIfNotAuthorized(Permissions.DocumentDelete);
+
+            IList<PimsProjectDocument> existingProjectDocuments = _projectRepository.GetAllByDocument(projectDocument.DocumentId);
+            if (existingProjectDocuments.Count == 1)
+            {
+                return await documentService.DeleteDocumentAsync(projectDocument.Document);
+            }
```
Comment (FuriousLlama) [inline source/backend/api/Services/DocumentFileService.cs:182]: This does not look correct. I think it should only delete the document if there are no other associations to the document, including other files, leases, etc. As it is, this one only checks if the document has a project document.

**Review by FuriousLlama — CHANGES_REQUESTED** [review]
The deletion of the document needs a second look

File: source/backend/api/Areas/Documents/DocumentRelationshipController.cs [hunk 2]
```diff
@@ -98,6 +99,10 @@ public IActionResult GetRelationshipDocuments(DocumentRelationType relationshipT
                     var leaseDocuments = _documentLeaseService.GetLeaseDocuments(parentId);
                     var mappedLeaseDocuments = _mapper.Map<List<DocumentRelationshipModel>>(leaseDocuments);
                     return new JsonResult(mappedLeaseDocuments);
+                case DocumentRelationType.Projects:
+                    var projectDocuments = _documentFileService.GetFileDocuments<PimsProjectDocument>(FileType.Project, parentId);
+                    var mappedprojectDocuments = _mapper.Map<List<DocumentRelationshipModel>>(projectDocuments);
+                    return new JsonResult(mappedprojectDocuments);
```
Comment (devinleighsmith) [inline source/backend/api/Areas/Documents/DocumentRelationshipController.cs:105]: nit: mappedProjectDocuments

File: source/backend/api/Areas/Documents/DocumentRelationshipController.cs [hunk 3]
```diff
@@ -120,26 +125,18 @@ public IActionResult GetRelationshipDocuments(DocumentRelationType relationshipT
             long parentId,
             [FromForm] DocumentUploadRequest uploadRequest)
         {
-            switch (relationshipType)
+            var response = relationshipType switch
```
Comment (devinleighsmith) [inline source/backend/api/Areas/Documents/DocumentRelationshipController.cs:128]: thanks for this, much more concise.

File: source/backend/api/Services/DocumentFileService.cs [hunk 4]
```diff
@@ -117,6 +123,35 @@ public async Task<DocumentUploadRelationshipResponse> UploadAcquisitionDocumentA
...
+            DocumentUploadResponse uploadResult = await documentService.UploadDocumentAsync(uploadRequest);
+
+            DocumentUploadRelationshipResponse relationshipResponse = new ()
+            {
+                UploadResponse = uploadResult,
+            };
+
+            if (uploadResult.Document.Id != 0)
+            {
+                // Create the pims document acquisition file relationship
```
Comment (devinleighsmith) [inline source/backend/api/Services/DocumentFileService.cs:140]: nit: comment copy/pasta

File: source/backend/dal/Repositories/ProjectRepository.cs [hunk 5]
```diff
@@ -124,6 +129,74 @@ public PimsProject Update(PimsProject project)
...
+                    .ThenInclude(d => d.DocumentStatusTypeCodeNavigation)
+                .Include(ad => ad.Document)
+                    .ThenInclude(d => d.DocumentType)
+                .Where(x => x.ProjectId == projectId)
+                .AsNoTracking()
+                .ToList();
+        }
+
+        /// <summary>
+        /// Get a list of all the document file relationships for a a given research file.
```
Comment (devinleighsmith) [inline source/backend/dal/Repositories/ProjectRepository.cs:170]: I don't think this comment is correct.

File: source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs [hunk 6]
```diff
@@ -109,6 +109,37 @@ public void GetPimsDocumentTypes_Acquisition_Success()
             acquisitionFileDocumentRepository.Verify(x => x.GetAllByAcquisitionFile(It.IsAny<long>()), Times.Once);
         }
 
+        [Fact]
+        public void GetPimsDocumentTypes_Project_Success()
```
Comment (devinleighsmith) [inline source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs:113]: why getPimsDocumentTypes? shouldn't this be GetAllProjectDocuments?

File: source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs [hunk 7]
```diff
@@ -418,5 +499,112 @@ public async void DeleteDocumentAcquisition_Success_Status_NotFound()
...
+
+            // Act
+            Func<Task> act = async () => await service.DeleteProjectDocumentAsync(doc);
+
+            // Assert
+            act.Should().Throw<NotAuthorizedException>();
+        }
+
+        [Fact]
+        [Trait("DocumentFileService", "Delete")]
```
Comment (devinleighsmith) [inline source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs:524]: nit: think the trait may be wrong here.

File: source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx [hunk 8]
```diff
@@ -103,6 +103,7 @@ export const DocumentUploadContainer: React.FunctionComponent<
   }, [props.relationshipType, retrieveDocumentTypes, isMounted, updateDocumentType]);
 
   const onUploadDocument = async (uploadRequest: Api_DocumentUploadRequest) => {
+    debugger;
```
Comment (devinleighsmith) [inline source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx:106]: @eddherrera this needs to be removed.

**Review by devinleighsmith — CHANGES_REQUESTED** [review]
debugger, as well as delete behaviour requires changing.

File: source/frontend/src/features/properties/map/project/ViewSelector.tsx [hunk 9]
```diff
@@ -38,7 +38,6 @@ export const ViewSelector = React.forwardRef<FormikProps<any>, IViewSelectorProp
         setProject={props.setProject}
         setContainerState={props.setContainerState}
         activeTab={props.activeTab}
-        isEditing={props.isEditing}
```
Comment (devinleighsmith) [inline source/frontend/src/features/properties/map/project/ViewSelector.tsx:41]: seems unrelated?

File: source/backend/api/Services/DocumentFileService.cs (hunk 1 above)
Comment (eddherrera) [reply source/backend/api/Services/DocumentFileService.cs]: Updated to check the total number of relationships instead of just one type.

File: source/backend/api/Areas/Documents/DocumentRelationshipController.cs (hunk 2 above)
Comment (eddherrera) [reply source/backend/api/Areas/Documents/DocumentRelationshipController.cs]: Updated

File: source/backend/api/Areas/Documents/DocumentRelationshipController.cs (hunk 3 above)
Comment (eddherrera) [reply source/backend/api/Areas/Documents/DocumentRelationshipController.cs:128]: 👍

File: source/backend/api/Services/DocumentFileService.cs (hunk 4 above)
Comment (eddherrera) [reply source/backend/api/Services/DocumentFileService.cs]: pasta removed

File: source/backend/dal/Repositories/ProjectRepository.cs (hunk 5 above)
Comment (eddherrera) [reply source/backend/dal/Repositories/ProjectRepository.cs]: Updated

File: source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs (hunk 6 above)
Comment (eddherrera) [reply source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs]: Updated

File: source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs (hunk 7 above)
Comment (eddherrera) [reply source/backend/tests/unit/api/Services/DocumentFileServiceTest.cs]: removed

File: source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx (hunk 8 above)
Comment (eddherrera) [reply source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx]: removed

File: source/frontend/src/features/properties/map/project/ViewSelector.tsx (hunk 9 above)
Comment (eddherrera) [reply source/frontend/src/features/properties/map/project/ViewSelector.tsx:41]: a little unrelated, but definitely not needed.

File: source/frontend/src/features/properties/map/project/detail/ProjectTabsContainer.test.tsx [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+// mock auth library
+jest.mock('@react-keycloak/web');
+
+const setProjectMock = jest.fn();
+const setEditModeMock = jest.fn();
+const setContainerStateMock = jest.fn();
+
+const mockProject = mockProjectGetResponse();
+
+describe('ResearchFileTabs component', () => {
```
[the remark this answers is not in the document]
Comment (eddherrera) [reply source/frontend/src/features/properties/map/project/detail/ProjectTabsContainer.test.tsx]: updated

**Review by devinleighsmith — APPROVED** [review]

**Comment by devinleighsmith** [comment]
@eddherrera please confirm that edit, view, delete, upload and download all work as expected in the new project context.

**Review by FuriousLlama — APPROVED** [review]
